gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java

```diff
@@ -211,9 +212,24 @@ private void handleRequest(final HttpServerRequest req, final Buffer bodyData, f
             cReq.headers().set("Authorization", "Basic " + base64UsernamePassword);
         }
 
-
         MultiMap headers = cReq.headers();
+        final String hostHeaderBefore = HttpHeaderUtil.getHeaderValue(headers, HOST_HEADER);
         final HeaderFunctions.EvalScope evalScope = rule.getHeaderFunction().apply(headers);
+        final String hostHeaderAfter = HttpHeaderUtil.getHeaderValue(headers, HOST_HEADER);
+        // see https://github.com/swisspush/gateleen/issues/394
+        if (hostHeaderAfter == null || hostHeaderAfter.equals(hostHeaderBefore)) {
```

Comment:
Is it possible to write a unit (or integration) test for the different scenarios?

Reply:
This will become very tricky as the handleRequest method is huge and must be split up to make a dedicated test just for this. Do you have any suggestions?

Reply:
Good point 👍 I think we definitively need tests which demonstrate the use-cases we want to cover with this feature. Because the code still lets tons of questions open for me.

BTW. [extract-method](https://www.martinfowler.com/bliki/FunctionLength.html) could help. Eg:
```java
EvalScope applyHeaderFunctions( HttpClientRequest req ){
    MultiMap headers = req.headers();
    // ...
    evalScope = rule.getHeaderFunction().apply( headers );
    // ...
    if( thereIsAReasonToDoStrangeThings() ){
        headers.set( HOST_HEADER, newHost );
    }
    // ...
    if( evalScopeHasErrors() ){
        return evalScope;
    }
    return null;
}
```

Reply:
That's exactly what I am working on actually. But it is not so simple ...